File: BookingNet/Domain.cs
namespace BookingNet;

public record Booking(string HotelName, string GuestName, DateTime CheckIn, DateTime CheckOut, Guid Id)
{
    public Booking(string hotelName, string guestName, DateTime checkIn, DateTime checkOut) : this(hotelName, guestName, checkIn, checkOut, Guid.NewGuid()) { }
}

File: BookingJava/src/main/java/pipelines/Domain.java
package pipelines;

import java.time.LocalDate;
import java.util.UUID;

record Booking(
        UUID id,
        String hotelName,
        String guestName,
        String email,
        LocalDate checkIn,
        LocalDate checkOut) {

    public Booking validate() {
        if (hotelName == null || hotelName.isBlank())
            throw new IllegalArgumentException("hotelName is required");

        if (guestName == null || guestName.isBlank())
            throw new IllegalArgumentException("guestName is required");

        if (email == null || email.isBlank() || !email.contains("@"))
            throw new IllegalArgumentException("A valid email is required");

        if (checkIn == null)
            throw new IllegalArgumentException("checkIn is required");
        if (checkOut == null)
            throw new IllegalArgumentException("checkOut is required");

        if (checkIn.isAfter(checkOut))
            throw new IllegalArgumentException("checkIn cannot be after checkOut");

        return this;
    }
}
